cleverhans/attacks.py

```diff
@@ -1,62 +1,233 @@
+
+
+class BasicIterativeMethod(Attack):
+    """
+    The Basic Iterative Method (Kurakin et al. 2016). The original paper used hard
+    labels for this attack; no label smoothing.
+    Paper link: https://arxiv.org/pdf/1607.02533.pdf
+    """
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None,
+                 clip_max=None, eps=0.3, eps_iter=0.05, ord='inf', nb_iter=10):
+        """
+        Create a BasicIterativeMethod instance.
+        :param eps: A float indicating the maximum allowed perturbation
+                    distance for each feature.
+        :param eps_iter: A float indicating the step size to use for each
+                        iteration of BIM (input variation parameter).
+        :param ord: A string indicating the norm order to use when computing
+                    gradients. This should be either 'inf', 'L1' or 'L2'.
+        :param nb_iter: An integer indicating the number of BIM iterations to run.
+        """
+        assert ord == 'inf' or ord == 'L1' or ord == 'L2'
+        super(BasicIterativeMethod, self).__init__(x, pred, y, backend,
+                                                   clip_min, clip_max)
+        self.eps = eps
+        self.eps_iter = eps_iter
+        self.nb_iter = nb_iter
+        self.fgm = FastGradientMethod(x, pred, y, backend, clip_min,
+                                      clip_max, eps_iter, ord)
+
+    def generate_numpy(self, X, Y=None, sess=None, batch_size=128):
+        """
+        Generate adversarial samples and return them in a Numpy array.
+        :param X: A Numpy array representing the feature matrix for the baseline
+                samples.
+        :param Y: A Numpy array representing the label matrix for the baseline
+                samples. Default is None (this is only used when using true labels
+                to craft adversarial samples)
+        :param sess: A TensorFlow session to use for evaluating the adversarial
+                samples (for 'tf' backend only). Default is None.
+        :param batch_size: An int indicating the batch size to use when evaluating
+                adversarial samples.
+        :return: A Numpy array holding the adversarial samples.
+        """
+        # verify that we are only using true labels if we indicated so previously
+        if Y is not None:
+            assert self.y is not None
+        super(BasicIterativeMethod, self).generate_numpy(sess)
+        upper_bound = X + self.eps
+        lower_bound = X - self.eps
+        X_adv = X
+        for i in range(self.nb_iter):
+            X_adv = self.fgm.generate_numpy(X_adv, Y, sess, batch_size)
+            X_adv = np.minimum(np.maximum(X_adv, lower_bound), upper_bound)
+
+        return X_adv
+
+
+class SaliencyMapMethod(Attack):
+    """
+    The Jacobian-based Saliency Map Method (Papernot et al. 2016).
+    Paper link: https://arxiv.org/pdf/1511.07528.pdf
+    """
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None,
```

Comment:
I may be missing something but I don't see where the Jacobian graph is defined here